File: CfxGenerator/CefApiDeclarations.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Public Class CefApiDeclarations

    Public ReadOnly ExportFunctions As CefExportFunction()
    Public ReadOnly CefStructTypes As CefStructType()
    Public ReadOnly CefEnumTypes As CefEnumType()
    Public ReadOnly StringCollectionTypes As StringCollectionType()
    Public ReadOnly StringCollectionFunctions As CefExportFunction()


    Public Sub New(exportFunctions As CefExportFunction(), cefStructTypes As CefStructType(), cefEnumTypes As CefEnumType(), stringCollectionTypes As StringCollectionType(), stringCollectionFunctions As CefExportFunction())
        Me.ExportFunctions = exportFunctions
        Me.CefStructTypes = cefStructTypes
        Me.CefEnumTypes = cefEnumTypes
        Me.StringCollectionTypes = stringCollectionTypes
        Me.StringCollectionFunctions = stringCollectionFunctions
    End Sub

    Private remoteFuncs As SortedDictionary(Of String, CefExportFunction)
    Private remoteStructs As SortedDictionary(Of String, CefStructType)

    Private retval As String()
    Public Function GetCfxApiFunctionNames() As String()
        If retval Is Nothing Then
            Dim list = New List(Of String)
            For Each f In ExportFunctions
                f.ApiIndex = list.Count
                list.Add(f.CfxName)
            Next
            For Each st In CefStructTypes
                For Each f In st.ClassBuilder.ExportFunctions
                    f.ApiIndex = list.Count
                    list.Add(f.CfxName)
                Next
                Select Case st.ClassBuilder.Category
                    Case StructCategory.ApiCalls
                        For Each sm In st.ClassBuilder.StructMembers
                            If sm.MemberType.IsCefCallbackType Then
                                sm.ApiIndex = list.Count
                                list.Add(sm.Callback.CfxApiFunctionName)
                            End If
                        Next
                    Case StructCategory.ApiCallbacks
                        st.ApiIndex = list.Count
                        list.Add(st.CfxName & "_ctor")
                        list.Add(st.CfxName & "_get_gc_handle")
                        list.Add(st.CfxName & "_set_managed_callback")
                    Case StructCategory.Values
                        st.ApiIndex = list.Count
                        list.Add(st.CfxName & "_ctor")
                        list.Add(st.CfxName & "_dtor")
                        For Each sm In st.ClassBuilder.StructMembers
                            If sm.Name <> "size" Then
                                sm.ApiIndex = list.Count
                                list.Add(st.CfxName & "_set_" & sm.Name)
                                list.Add(st.CfxName & "_get_" & sm.Name)
                            End If
                        Next
                End Select
            Next

            For Each f In StringCollectionFunctions
                f.ApiIndex = list.Count
                list.Add(f.CfxName)
            Next

            retval = list.ToArray()
        End If
        Return retval
    End Function

    Public Function GetRemoteDeclarations() As CefApiDeclarations

        remoteFuncs = New SortedDictionary(Of String, CefExportFunction)
        remoteStructs = New SortedDictionary(Of String, CefStructType)

        For Each f In ExportFunctions
            If Not GeneratorConfig.IsBrowserProcessOnly(f.Name) Then
                AddRemoteFunc(f)
            End If
        Next

        Return New CefApiDeclarations(remoteFuncs.Values.ToArray(), remoteStructs.Values.ToArray(), CefEnumTypes, Nothing, Nothing)

    End Function

    Private Sub AddRemoteFunc(f As CefExportFunction)
        If remoteFuncs.ContainsKey(f.Name) Then Return
        remoteFuncs.Add(f.Name, f)
        AnalyzeSignature(f.Signature)
    End Sub

    Private Sub AnalyzeSignature(sig As Signature)
        If sig.PublicReturnType.IsCefStructType Then
            AddRemoteStruct(sig.PublicReturnType.AsCefStructType)
        ElseIf sig.PublicReturnType.IsCefStructPtrType Then
            AddRemoteStruct(sig.PublicReturnType.AsCefStructPtrType.Struct)
        End If

        For Each arg In sig.ManagedArguments
            If arg.ArgumentType.IsCefStructType Then
                AddRemoteStruct(arg.ArgumentType.AsCefStructType)
            ElseIf arg.ArgumentType.IsCefStructPtrType Then
                AddRemoteStruct(arg.ArgumentType.AsCefStructPtrType.Struct)
            End If
        Next
    End Sub

    Private Sub AddRemoteStruct(s As CefStructType)


        If remoteStructs.ContainsKey(s.Name) Then Return
        remoteStructs.Add(s.Name, s)

        For Each f In s.ClassBuilder.ExportFunctions
            If Not GeneratorConfig.IsBrowserProcessOnly(f.Name) Then
                AnalyzeSignature(f.Signature)
            End If
        Next

        For Each sm In s.ClassBuilder.StructMembers
            If Not GeneratorConfig.IsBrowserProcessOnly(s.Name & "::" & sm.Name) Then
                If sm.MemberType.IsCefStructType Then
                    AddRemoteStruct(sm.MemberType.AsCefStructType)
                ElseIf sm.MemberType.IsCefStructPtrType Then
                    AddRemoteStruct(sm.MemberType.AsCefStructPtrType.Struct)
                ElseIf sm.MemberType.IsCefCallbackType Then
                    AnalyzeSignature(sm.MemberType.AsCefCallbackType.Signature)
                End If
            End If
        Next

    End Sub


    Private funcs As CefExportFunction()
    Public Function AllExportFunctions() As CefExportFunction()
        If funcs Is Nothing Then
            Dim ff = New List(Of CefExportFunction)(ExportFunctions)
            For Each t In CefStructTypes
                ff.AddRange(t.ClassBuilder.ExportFunctions)
            Next
            funcs = ff.ToArray()
        End If
        Return funcs
    End Function

End Class

File: CfxGenerator/CefApiDeclarations.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

public class CefApiDeclarations {
    public readonly CefExportFunction[] ExportFunctions;
    public readonly CefStructType[] CefStructTypes;
    public readonly CefEnumType[] CefEnumTypes;
    public readonly StringCollectionType[] StringCollectionTypes;

    public readonly CefExportFunction[] StringCollectionFunctions;

    public CefApiDeclarations(CefExportFunction[] exportFunctions, CefStructType[] cefStructTypes, CefEnumType[] cefEnumTypes, StringCollectionType[] stringCollectionTypes, CefExportFunction[] stringCollectionFunctions) {
        this.ExportFunctions = exportFunctions;
        this.CefStructTypes = cefStructTypes;
        this.CefEnumTypes = cefEnumTypes;
        this.StringCollectionTypes = stringCollectionTypes;
        this.StringCollectionFunctions = stringCollectionFunctions;
    }

    private SortedDictionary<string, CefExportFunction> remoteFuncs;

    private SortedDictionary<string, CefStructType> remoteStructs;
    private string[] retval;

    public string[] GetCfxApiFunctionNames() {
        if(retval == null) {
            var list = new List<string>();
            foreach(var f in ExportFunctions) {
                f.ApiIndex = list.Count;
                list.Add(f.CfxName);
            }
            foreach(var st in CefStructTypes) {
                if(st.ClassBuilder.ExportFunctions != null) {
                    foreach(var f in st.ClassBuilder.ExportFunctions) {
                        f.ApiIndex = list.Count;
                        list.Add(f.CfxName);
                    }
                }
                switch(st.ClassBuilder.Category) {
                    case StructCategory.ApiCalls:
                        foreach(var cb in st.ClassBuilder.CallbackFunctions) {
                            cb.ApiIndex = list.Count;
                            list.Add(cb.CfxApiFunctionName);
                        }

                        break;

                    case StructCategory.ApiCallbacks:
                        st.ApiIndex = list.Count;
                        list.Add(st.CfxName + "_ctor");
                        list.Add(st.CfxName + "_get_gc_handle");
                        list.Add(st.CfxName + "_set_callback");
                        break;

                    case StructCategory.Values:
                        st.ApiIndex = list.Count;
                        list.Add(st.CfxName + "_ctor");
                        list.Add(st.CfxName + "_dtor");
                        foreach(var sm in st.ClassBuilder.StructMembers) {
                            sm.ApiIndex = list.Count;
                            list.Add(st.CfxName + "_set_" + sm.Name);
                            list.Add(st.CfxName + "_get_" + sm.Name);
                        }

                        break;
                }
            }

            foreach(var f in StringCollectionFunctions) {
                f.ApiIndex = list.Count;
                list.Add(f.CfxName);
            }

            retval = list.ToArray();
        }
        return retval;
    }

    public CefApiDeclarations GetRemoteDeclarations() {
        remoteFuncs = new SortedDictionary<string, CefExportFunction>();
        remoteStructs = new SortedDictionary<string, CefStructType>();

        foreach(var f in ExportFunctions) {
            if(!GeneratorConfig.IsBrowserProcessOnly(f.Name)) {
                AddRemoteFunc(f);
            }
        }

        foreach(var t in CefStructTypes) {
            if(t.ClassBuilder.ExportFunctions != null) {
                foreach(var f in t.ClassBuilder.ExportFunctions) {
                    if(!GeneratorConfig.IsBrowserProcessOnly(f.Name)) {
                        AnalyzeSignature(f.Signature);
                    }
                }
            }
        }

        return new CefApiDeclarations(remoteFuncs.Values.ToArray(), remoteStructs.Values.ToArray(), CefEnumTypes, null, null);
    }

    private void AddRemoteFunc(CefExportFunction f) {
        if(remoteFuncs.ContainsKey(f.Name))
            return;
        remoteFuncs.Add(f.Name, f);
        AnalyzeSignature(f.Signature);
    }

    private void AnalyzeSignature(Signature sig) {
        if(sig.PublicReturnType.IsCefStructType) {
            AddRemoteStruct(sig.PublicReturnType.AsCefStructType);
        } else if(sig.PublicReturnType.IsCefStructPtrType) {
            AddRemoteStruct(sig.PublicReturnType.AsCefStructPtrType.Struct);
        }

        foreach(var arg in sig.ManagedParameters) {
            if(arg.ParameterType.IsCefStructType) {
                AddRemoteStruct(arg.ParameterType.AsCefStructType);
            } else if(arg.ParameterType.IsCefStructPtrType) {
                AddRemoteStruct(arg.ParameterType.AsCefStructPtrType.Struct);
            }
        }
    }

    private void AddRemoteStruct(CefStructType s) {
        if(remoteStructs.ContainsKey(s.Name))
            return;
        remoteStructs.Add(s.Name, s);

        switch(s.ClassBuilder.Category) {
            case StructCategory.Values:
                foreach(var sm in s.ClassBuilder.StructMembers) {
                    if(!GeneratorConfig.IsBrowserProcessOnly(s.Name + "::" + sm.Name)) {
                        if(sm.MemberType.IsCefStructType) {
                            AddRemoteStruct(sm.MemberType.AsCefStructType);
                        } else if(sm.MemberType.IsCefStructPtrType) {
                            AddRemoteStruct(sm.MemberType.AsCefStructPtrType.Struct);
                        }
                    }
                }
                break;
            case StructCategory.ApiCalls:
                foreach(var f in s.ClassBuilder.ExportFunctions) {
                    if(!GeneratorConfig.IsBrowserProcessOnly(f.Name)) {
                        AnalyzeSignature(f.Signature);
                    }
                }
                goto case StructCategory.ApiCallbacks;
            case StructCategory.ApiCallbacks:
                foreach(var cb in s.ClassBuilder.CallbackFunctions) {
                    if(!GeneratorConfig.IsBrowserProcessOnly(s.Name + "::" + cb.Name)) {
                        AnalyzeSignature(cb.Signature);
                    }
                }
                break;
        }
    }

    private CefExportFunction[] funcs;

    public CefExportFunction[] AllExportFunctions() {
        if(funcs == null) {
            var ff = new List<CefExportFunction>(ExportFunctions);
            foreach(var t in CefStructTypes) {
                if(t.ClassBuilder.ExportFunctions != null)
                    ff.AddRange(t.ClassBuilder.ExportFunctions);
            }
            funcs = ff.ToArray();
        }
        return funcs;
    }
}